File: src/CSharp/SodaFlow.Tests/PerformanceTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using TUnit.Assertions;
using TUnit.Assertions.Enums;
using TUnit.Assertions.Extensions;
using TUnit.Core;
using SodaFlow.Functional;

namespace SodaFlow.Tests;

public class PerformanceTests
{
    [Test]
    public async Task TestMerge()
    {
        StreamSink<Unit> s = Stream.CreateSink<Unit>();

        TestObject[] obj =
            Transaction.Run(() =>
            {
                StreamLoop<bool> loop = Stream.CreateLoop<bool>();
                CellStreamSink<int> s1 = Cell.CreateStreamSink<int>();
                CellStreamSink<int> s2 = Cell.CreateStreamSink<int>();

                TestObject[] l =
                [
                    .. Enumerable.Range(start: 0, count: 5000)
                        .Select(_ => new TestObject(s: loop, s1: s1, s2: s2))
                ];

                loop.Loop(s.Snapshot(l.Select(static o => o.Cell).Lift()).Map(static o => o.All(static v => v == 0)));
                return l;
            });

        int[] values = [.. obj.Select(static v => v.CurrentValue)];

        await Assert.That(values).IsEquivalentTo(Enumerable.Range(start: 1, count: 5000).Select(static _ => 0), CollectionOrdering.Matching);
    }

    private sealed class TestObject
    {
        // ReSharper disable once NotAccessedField.Local
        private readonly IListener l;
        private Lazy<int> currentValue = new(static () => 0);

        public TestObject(Stream<bool> s, Stream<int> s1, Stream<int> s2)
        {
            (Cell<int> cell, IStrongListener l) =
                Transaction.Run(() =>
                {
                    Cell<int> cellLocal = s.Map(static v => v ? 1 : 0).OrElse(s1).OrElse(s2).Hold(0);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell2 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell3 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell4 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell5 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell6 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell7 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell8 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    // ReSharper disable once UnusedVariable - Keeps Cell in scope.
                    Cell<int> cell9 =
                        s1.Snapshot(c: cellLocal, f: static (left, right) => left + right)
                            .Filter(static v => v > 5)
                            .OrElse(
                                s.Snapshot(s1.Hold(0).Lift(c2: s2.Hold(1), f: static (left, right) => left + right))
                                    .Map(static v => v + 1))
                            .Hold(3);

                    this.currentValue = cellLocal.SampleLazy();
                    IStrongListener lLocal = cellLocal.Updates().ListenStrong(v => this.CurrentValue = v);

                    return (cellLocal, lLocal);
                });

            this.Cell = cell;
            this.l = l;
        }

        public Cell<int> Cell { get; }

        public int CurrentValue
        {
            get => this.currentValue.Value;
            private set => this.currentValue = new Lazy<int>(() => value);
        }
    }

    [Test]
    public async Task TestRunConstruct()
    {
        CellSink<IReadOnlyList<TestObject2>> objects =
            Transaction.Run(static () =>
            {
                IReadOnlyList<TestObject2> o2 =
                [
                    .. Enumerable.Range(start: 0, count: 10000)
                        .Select(static n =>
                            new TestObject2(
                                initialIsSelected: n < 1500,
                                selectAllStream: Stream.Never<bool>()))
                ];

                CellSink<IReadOnlyList<TestObject2>> objectsLocal = Cell.CreateSink(o2);

                return objectsLocal;
            });

        Transaction.Run(() =>
        {
            objects.Send(
            [
                .. Enumerable.Range(start: 0, count: 20000)
                    .Select(static n =>
                        new TestObject2(initialIsSelected: n < 500, selectAllStream: Stream.Never<bool>()))
            ]);

            return Unit.Value;
        });
    }

    [Test]
    public async Task TestRunConstruct2()
    {
        (var objectsAndIsSelected, Stream<bool> selectAllStream, CellSink<IReadOnlyList<TestObject2>> objects) =
            Transaction.Run(static () =>
            {
                CellLoop<bool?> allSelectedCellLoop = Cell.CreateLoop<bool?>();
                StreamSink<Unit> toggleAllSelectedStreamLocal = Stream.CreateSink<Unit>();

                Stream<bool> selectAllStreamLocal =
                    toggleAllSelectedStreamLocal.Snapshot(allSelectedCellLoop).Map(static a => a != true);

                IReadOnlyList<TestObject2> o2 =
                [
                    .. Enumerable.Range(start: 0, count: 10000)
                        .Select(n =>
                            new TestObject2(
                                initialIsSelected: n < 1500,
                                selectAllStream: selectAllStreamLocal))
                ];

                CellSink<IReadOnlyList<TestObject2>> objectsLocal = Cell.CreateSink(o2);

                var objectsAndIsSelectedLocal =
                    objectsLocal
                        .Map(static oo =>
                            oo.Select(static o => o.IsSelected.Map(s => new { Object = o, IsSelected = s })).Lift())
                        .SwitchC();

                bool defaultValue = o2.Count < 1;

                Cell<bool?> allSelected =
                    objectsAndIsSelectedLocal.Map(oo =>
                        !oo.Any()
                            ? defaultValue
                            : oo.All(static o => o.IsSelected)
                                ? true
                                : oo.All(static o => !o.IsSelected)
                                    ? (bool?)false
                                    : null);

                allSelectedCellLoop.Loop(allSelected);

                return (objectsAndIsSelectedLocal, selectAllStreamLocal, objectsLocal);
            });

        List<int> @out = [];

        using (Transaction.Run(() =>
                   objectsAndIsSelected.Map(static oo => oo.Count(static o => o.IsSelected))
                       .Values()
                       .ListenStrong(@out.Add)))
        {
            Transaction.Run(() =>
            {
                objects.Send(
                [
                    .. Enumerable.Range(start: 0, count: 20000)
                        .Select(n =>
                            new TestObject2(initialIsSelected: n < 500, selectAllStream: selectAllStream))
                ]);

                return Unit.Value;
            });
        }

        await Assert.That(@out).IsEquivalentTo(new[] { 1500, 500 }, CollectionOrdering.Matching);
    }

    private sealed class TestObject2
    {
        public TestObject2(bool initialIsSelected, Stream<bool> selectAllStream)
        {
            this.IsSelectedStreamSink = Stream.CreateSink<bool>();
            this.IsSelected = selectAllStream.OrElse(this.IsSelectedStreamSink).Hold(initialIsSelected);
        }

        private StreamSink<bool> IsSelectedStreamSink { get; }
        public Cell<bool> IsSelected { get; }
    }
}

File: src/FSharp/SodaFlow.FSharp.Tests/PerformanceTests.fs
module SodaFlow.Tests.Performance

open System.Collections.Generic
open NUnit.Framework
open SodaFlow

[<AutoOpen>]
module private Types =

    type TestObject(s: Stream<bool>, s1: Stream<int>, s2: Stream<int>) as this =
        let mutable currentValue = lazy 0

        let cell, l =
            runT (fun () ->
                let cell =
                    ((s |> mapS (fun v -> if v then 1 else 0), s1) |> orElseS, s2)
                    |> orElseS
                    |> holdS 0

                let createCell () =
                    (s1 |> snapshotC cell (+) |> filterS (flip (>) 5),
                     s
                     |> snapshotAndTakeC ((s1 |> holdS 0, s2 |> holdS 1) |> lift2C (+))
                     |> mapS ((+) 1))
                    |> orElseS
                    |> holdS 3

                let _cell2 = createCell ()
                let _cell3 = createCell ()
                let _cell4 = createCell ()
                let _cell5 = createCell ()
                let _cell6 = createCell ()
                let _cell7 = createCell ()
                let _cell8 = createCell ()
                let _cell9 = createCell ()
                currentValue <- cell |> sampleLazyC
                let l = cell |> updatesC |> listenStrongS (fun v -> this.CurrentValue <- v)
                (cell, l))

        member private _.L = l
        member _.Cell = cell

        member _.CurrentValue
            with get () = currentValue.Value
            and set value = currentValue <- lazy value

    type TestObject2(id: int, initialIsSelected: bool, selectAllStream: Stream<bool>) =
        let isSelectedStreamSink = sinkS ()

        let isSelected =
            (selectAllStream, isSelectedStreamSink) |> orElseS |> holdS initialIsSelected

        member _.Id = id
        member _.IsSelectedStreamSink = isSelectedStreamSink
        member _.IsSelected = isSelected

[<TestFixture>]
type ``Performance Tests``() =

    [<Test>]
    member _.``Test Merge``() =
        let s = sinkS<unit> ()

        let struct (_, obj) =
            loopS (fun loop ->
                let s1 = sinkCS ()
                let s2 = sinkCS ()
                let l = Array.init 5000 (fun _ -> TestObject(loop, s1, s2))

                struct (s
                        |> snapshotAndTakeC (l |> Seq.map (fun o -> o.Cell) |> liftAllC id)
                        |> mapS (Seq.forall (fun v -> v = 0)),
                        l))

        let values = obj |> Array.map (fun o -> o.CurrentValue)
        CollectionAssert.AreEqual(Seq.init 5000 (fun _ -> 0), values)

    [<Test>]
    member _.``Test Run Construct``() =
        let objects =
            runT (fun () ->
                let o2 = List.init 10000 (fun n -> TestObject2(n, n < 1500, neverS ()))
                sinkC o2)

        runT (fun () -> objects |> sendC (List.init 20000 (fun n -> TestObject2(n, n < 500, neverS ()))))

    [<Test>]
    member _.``Test Run Construct 2``() =
        let struct (_, (objectsAndIsSelected, selectAllStream, objects)) =
            loopC (fun allSelected ->
                let toggleAllSelectedStream = sinkS ()

                let selectAllStream =
                    toggleAllSelectedStream
                    |> snapshotAndTakeC allSelected
                    |> mapS (fun a ->
                        match a with
                        | Some a -> not a
                        | None -> true)

                let o2 = List.init 10000 (fun n -> TestObject2(n, n < 1500, selectAllStream))
                let objects = sinkC o2

                let objectsAndIsSelected =
                    objects
                    |> mapC (Seq.map (fun o -> o.IsSelected |> mapC (fun s -> (o, s))) >> liftAllC id)
                    |> switchC

                let defaultValue = Some(o2.Length < 1)

                let allSelected =
                    objectsAndIsSelected
                    |> mapC (fun oo ->
                        if oo.Count > 0 then
                            (if oo |> Seq.forall snd then
                                 Some true
                             else
                                 (if oo |> Seq.forall (fun (_, isSelected) -> not isSelected) then
                                      Some false
                                  else
                                      None))
                        else
                            defaultValue)

                struct (allSelected, (objectsAndIsSelected, selectAllStream, objects)))

        let out = List<_>()

        (use _l =
            runT (fun () ->
                objectsAndIsSelected
                |> mapC (Seq.where snd >> Seq.length)
                |> listenStrongC out.Add)

         runT (fun () ->
             objects
             |> sendC (List.init 20000 (fun n -> TestObject2(n, n < 500, selectAllStream)))))

        CollectionAssert.AreEqual([ 1500; 500 ], out)
